packages/pages/src/dev/server/middleware/sendAppHTML.ts

```diff
@@ -30,28 +34,40 @@ export default async function sendAppHTML(
     return;
   }
 
-  const appHtml = ReactDOMServer.renderToString(
-    React.createElement(templateModuleInternal.default || "", props)
+  const clientServerRenderTemplates = getGlobalClientServerRenderTemplates(
+    projectStructure.templatesRoot,
+    projectStructure.scopedTemplatesPath
   );
 
+  const serverRenderTemplateModule = (await vite.ssrLoadModule(
+    clientServerRenderTemplates.serverRenderTemplatePath
+  )) as RenderTemplate;
+
+  const serverHtml = await serverRenderTemplateModule.render({
+    Page: templateModuleInternal.default!,
+    pageProps: props,
+  });
+
   const headConfig = templateModuleInternal.getHeadConfig
     ? templateModuleInternal.getHeadConfig(props)
     : undefined;
 
-  const template = await vite.transformIndexHtml(pathname, templateBase);
-
-  // Inject the app-rendered HTML into the template. Only invoke the users headFunction
-  // if they are rendering by way of a default export and not a custom render function.
-  const html = template.replace(`<!--app-html-->`, appHtml).replace(
-    `<!--app-head-->`,
-    `<head>
-        <script type="text/javascript">
-          window._RSS_PROPS_ = ${JSON.stringify(props)};
-          window._RSS_TEMPLATE_PATH_ = '${templateModuleInternal.path}';
-          window._RSS_LANG_ = '${getLang(headConfig, props)}';
-        </script>
-        ${headConfig ? renderHeadConfigToString(headConfig) : ""}
-      </head>`
+  const clientHtml = getHydrationTemplate(
```

Comment:
I'm still unclear here why we need to get the client and server side templates here. shouldn't we just have to get the client html?

Reply:
We need the server template too because it defines how to render things:

```
const { Page, pageProps } = pageContext;
  
const viewHtml = ReactDOMServer.renderToString(
    <Page {...pageProps} /></CacheProvider>
);

return `<!DOCTYPE html>
    <html lang="${language}">
      <head>whatever here</head>
      <body>
        <div>Duval is the GOAT!!!</div>
        <div id="test">${viewHtml}</div>
      </body>
    </html>`;
```

We can't hardcode this because then:
1. a user isn't in full control
2. the target id of the div needs to match between the client and server
3. the original impetus beside full customizability was for Heartland to be able to use an Emotion cache, which requires wrapping things on both sides like

```
_server.tsx
const viewHtml = ReactDOMServer.renderToString(
    <CacheProvider value={cache}>
      <Page {...pageProps} />
    </CacheProvider>
  );

_client.tsx
ReactDOM.hydrate(
    <CacheProvider value={cache}>
      <Page {...pageProps} />
    </CacheProvider>,
    document.getElementById("test")
  );
```

Reply:
Okay that makes sense. thanks